src/connector/builder.rs

```diff
@@ -0,0 +1,234 @@
+    // ALPN is built piecemeal without the need to read back this field
+    #[allow(dead_code)]
+    enable_http1: bool,
+}
+
+impl ConnectorBuilder<WantsTlsConfig> {
+    /// Creates a new [ConnectorBuilder]
+    pub fn new() -> Self {
+        Self(WantsTlsConfig(()))
+    }
+
+    /// Passes a rustls [ClientConfig] to configure the TLS connection
+    ///
+    /// The [alpn_protocols](ClientConfig::alpn_protocols) field will be rewritten to
+    /// match the enabled schemes (see
+    /// [enable_http1](ConnectorBuilder::enable_http1),
+    /// [enable_http2](ConnectorBuilder::enable_http2)) before the
+    /// connector is built.
+    pub fn with_tls_config(self, config: ClientConfig) -> ConnectorBuilder<WantsSchemes> {
+        ConnectorBuilder(WantsSchemes { tls_config: config })
+    }
+
+    /// Shorthand for using rustls' [safe defaults][with_safe_defaults]
+    /// and native roots
+    ///
+    /// See [ConfigBuilderExt::with_native_roots]
+    ///
+    /// [with_safe_defaults]: rustls::ConfigBuilder::with_safe_defaults
+    #[cfg(feature = "rustls-native-certs")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-native-certs")))]
+    pub fn with_native_roots(self) -> ConnectorBuilder<WantsSchemes> {
+        self.with_tls_config(
+            ClientConfig::builder()
+                .with_safe_defaults()
+                .with_native_roots(),
+        )
+    }
+
+    /// Shorthand for using rustls' [safe defaults][with_safe_defaults]
+    /// and Mozilla roots
+    ///
+    /// See [ConfigBuilderExt::with_webpki_roots]
+    ///
+    /// [with_safe_defaults]: rustls::ConfigBuilder::with_safe_defaults
+    #[cfg(feature = "webpki-roots")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "webpki-roots")))]
+    pub fn with_webpki_roots(self) -> ConnectorBuilder<WantsSchemes> {
+        self.with_tls_config(
+            ClientConfig::builder()
+                .with_safe_defaults()
+                .with_webpki_roots(),
+        )
+    }
+}
+
+impl Default for ConnectorBuilder<WantsTlsConfig> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ConnectorBuilder<WantsSchemes> {
+    /// Enforce the use of HTTPS when connecting
+    ///
+    /// Only URLs using the HTTPS scheme will be connectable.
+    pub fn https_only(self) -> ConnectorBuilder<WantsProtocols1> {
+        ConnectorBuilder(WantsProtocols1 {
+            tls_config: self.0.tls_config,
+            https_only: true,
+        })
+    }
+
+    /// Allow both HTTPS and HTTP when connecting
+    ///
+    /// HTTPS URLs will be handled through rustls,
+    /// HTTP URLs will be handled by the lower-level connector.
+    pub fn https_or_http(self) -> ConnectorBuilder<WantsProtocols1> {
+        ConnectorBuilder(WantsProtocols1 {
+            tls_config: self.0.tls_config,
+            https_only: false,
+        })
+    }
+}
+
+impl WantsProtocols1 {
+    fn wrap_connector<H>(mut self, conn: H) -> HttpsConnector<H> {
+        self.tls_config.alpn_protocols.clear();
```

Comment:
Since the connector will be speaking the protocols, it makes sense to keep alpn_protocols completely in sync with what we want to handle.

(There's also some subtleties with order being significant and with an empty alpn_protocols meaning "http/1.1"; building what we want makes it easier to have everything consistent)